File: src-tauri/src/envelope.rs
use serde_json::{json, Value};

use crate::error::AppResult;

/// PEYT 信封协议 (见 docs/superpowers/specs/2026-08-04-pure-json-envelope-design.md)
///
/// 载体: view_type=Text 的普通消息, 正文就是纯 JSON, 无前缀。
/// 结构:
///   {
///     "type": "text",        // 注册表判别符(聊天消息类型), 未知则显示原文
///     "id": "<uuid>",        // 发送端幂等键, 防重试/多端重复处理
///     "payload": { "text": "你好" }  // 类型专属载荷; 所有 type 都有 text 字段填消息体正文
///   }
/// 无 version / 无 from: 兼容性靠 payload 强校验, 发送者取消息 from_id。
/// 无 timestamp: core 的 msgs.timestamp 已提供, 信封不重复携带。
/// 字段全名风格(不缩写)——消息会出现在聊天流里,可读性优先。

/// 构建信封字符串: 纯 JSON。
pub fn build_envelope(type_: &str, payload: Value) -> AppResult<String> {
    let id = uuid::Uuid::new_v4().to_string();
    let env = json!({
        "type": type_,
        "id": id,
        "payload": payload,
    });
    Ok(env.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn envelope_structure() {
        let s = build_envelope("text", json!({"text": "你好"})).unwrap();
        // 纯 JSON, 无前缀
        assert!(s.starts_with('{'));
        let json: Value = serde_json::from_str(&s).unwrap();
        assert_eq!(json["type"], "text");
        assert!(json["id"].as_str().unwrap().len() == 36); // uuid
        assert_eq!(json["payload"]["text"], "你好");
        // 无 version / from / timestamp 字段
        assert!(json.get("version").is_none());
        assert!(json.get("from").is_none());
        assert!(json.get("timestamp").is_none());
    }

    #[test]
    fn envelope_text_with_markdown() {
        let s = build_envelope("text", json!({"text": "**hi**", "markdown": true})).unwrap();
        let json: Value = serde_json::from_str(&s).unwrap();
        assert_eq!(json["payload"]["markdown"], true);
        assert_eq!(json["payload"]["text"], "**hi**");
    }
}

File: src/utils/envelope.ts
// 纯 JSON 信封协议解析(见 docs/superpowers/specs/2026-08-04-pure-json-envelope-design.md)
// 信封是 view_type=Text 的普通消息, 正文就是纯 JSON:
//   { "type": <string>, "id": <uuid>, "payload": { "text": <string>, ... } }
// 约定: 所有 type 的 payload 都带 text 字段, 填充消息体正文。
// 识别靠形状启发式: 解析成功 + type/id/payload 三字段齐全 → 信封; 否则 → 普通文本。

export interface Envelope {
  type: string;
  id: string;
  payload: Record<string, unknown> & { text?: unknown };
}

/** 尝试解析信封。非信封 / 结构不合法 → null(调用方显示原文)。 */
export function tryParseEnvelope(text: string): Envelope | null {
  if (!text || text[0] !== '{') return null;
  let obj: unknown;
  try {
    obj = JSON.parse(text);
  } catch {
    return null;
  }
  if (typeof obj !== 'object' || obj === null) return null;
  const e = obj as Record<string, unknown>;
  if (typeof e.type !== 'string') return null;
  if (typeof e.id !== 'string') return null;
  if (typeof e.payload !== 'object' || e.payload === null) return null;
  return {
    type: e.type,
    id: e.id,
    payload: e.payload as Envelope['payload'],
  };
}

/** 取信封正文文本(payload.text)。缺失或非字符串 → null。 */
export function envelopeText(env: Envelope): string | null {
  const t = env.payload.text;
  return typeof t === 'string' ? t : null;
}

/** 取信封 md 标记:payload.markdown === true 才 true(布尔严格校验)。 */
export function envelopeMarkdown(env: Envelope): boolean {
  return env.payload.markdown === true;
}

/** 还原消息正文: 是信封 → payload.text; 否则 → 原文。 */
export function resolveMessageText(text: string): string {
  const env = tryParseEnvelope(text);
  if (!env) return text;
  return envelopeText(env) ?? text;
}
